File: services/ordering/src/main/java/io/wkrzywiec/fooddelivery/ordering/domain/OrderingEventMapper.java
package io.wkrzywiec.fooddelivery.ordering.domain;

import io.wkrzywiec.fooddelivery.commons.event.IntegrationEventMapper;
import io.wkrzywiec.fooddelivery.commons.event.IntegrationMessageBody;
import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent;
import io.wkrzywiec.fooddelivery.ordering.domain.outgoing.*;
import org.mapstruct.Mapper;
import org.mapstruct.factory.Mappers;

import static java.lang.String.format;

@Mapper
public interface OrderingEventMapper extends IntegrationEventMapper {

    OrderingEventMapper INSTANCE = Mappers.getMapper( OrderingEventMapper.class );

    default IntegrationMessageBody map(DomainEvent domainEvent) {
        return switch (domainEvent) {
            case OrderingEvent.OrderCreated de -> map(de);
            case OrderingEvent.OrderCanceled de -> map(de);
            case OrderingEvent.OrderInProgress de -> map(de);
            case OrderingEvent.TipAddedToOrder de -> map(de);
            case OrderingEvent.OrderCompleted de -> map(de);

            default -> throw new IllegalArgumentException(format("Failed to map DomainEvent to IntegrationMessageBody. Unknown DomainEvent: %s", domainEvent.getClass()));
        };
    }

    OrderCreated map(OrderingEvent.OrderCreated created);
    OrderCanceled map(OrderingEvent.OrderCanceled canceled);
    OrderInProgress map(OrderingEvent.OrderInProgress inProgress);
    TipAddedToOrder map(OrderingEvent.TipAddedToOrder tipAdded);
    OrderCompleted map(OrderingEvent.OrderCompleted completed);
}

File: services/ordering/src/main/kotlin/io/wkrzywiec/fooddelivery/ordering/domain/OrderingEventMapper.kt
package io.wkrzywiec.fooddelivery.ordering.domain

import io.wkrzywiec.fooddelivery.commons.event.IntegrationEventMapper
import io.wkrzywiec.fooddelivery.commons.event.IntegrationMessageBody
import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent
import io.wkrzywiec.fooddelivery.ordering.domain.outgoing.*
import org.mapstruct.Mapper
import org.mapstruct.Mapping

@Mapper
@JvmDefaultWithCompatibility
interface OrderingEventMapper : IntegrationEventMapper {

    override fun map(domainEvent: DomainEvent): IntegrationMessageBody {
        return when (domainEvent) {
            is OrderingEvent.OrderCreated -> map(domainEvent)
            is OrderingEvent.OrderCanceled -> map(domainEvent)
            is OrderingEvent.OrderInProgress  -> map(domainEvent)
            is OrderingEvent.TipAddedToOrder -> map(domainEvent)
            is OrderingEvent.OrderCompleted -> map(domainEvent)

            else -> throw IllegalArgumentException("Failed to map DomainEvent to IntegrationMessageBody. Unknown DomainEvent: ${domainEvent::class.simpleName}")
        }
    }

    @Mapping(source = "orderId", target = "id")
    @Mapping(source = "version", target = "aggregateVersion")
    fun map(created: OrderingEvent.OrderCreated): OrderCreated

    @Mapping(source = "orderId", target = "id")
    @Mapping(source = "version", target = "aggregateVersion")
    fun map(canceled: OrderingEvent.OrderCanceled): OrderCanceled

    @Mapping(source = "orderId", target = "id")
    @Mapping(source = "version", target = "aggregateVersion")
    fun map(inProgress: OrderingEvent.OrderInProgress): OrderInProgress

    @Mapping(source = "orderId", target = "id")
    @Mapping(source = "version", target = "aggregateVersion")
    fun map(tipAdded: OrderingEvent.TipAddedToOrder): TipAddedToOrder

    @Mapping(source = "orderId", target = "id")
    @Mapping(source = "version", target = "aggregateVersion")
    fun map(completed: OrderingEvent.OrderCompleted): OrderCompleted
}